Fabric.Terminology.SqlServer/Persistence/SqlClientTermValueSetRepository.UoWOperation.cs

```diff
@@ -0,0 +1,138 @@
+
+    internal partial class SqlClientTermValueSetRepository
+    {
+        private static ValueSetCodeDto ConvertToValueSetCodeDto(Guid valueSetGuid, ICodeSystemCode code) =>
+            new ValueSetCodeDto(code) { ValueSetGUID = valueSetGuid };
+
+        private AddRemoveCodeOperations PrepareAddRemoveCodes(
+            Guid valueSetGuid,
+            IReadOnlyCollection<ICodeSystemCode> codesToAdd,
+            IReadOnlyCollection<ICodeSystemCode> codesToRemove)
+        {
+            var currentCodeDtos = this.uowManager.GetCodeDtos(valueSetGuid);
+
+            //// There could be duplicates between the codesToAdd and codesToRemove collections.
+            //// In cases where a code is found in both lists, it will simply be removed from each list
+            //// since we are using SqlBulkCopy to batch insert records the "added" codes (e.g. there is no
+            //// update/delete capability in that process and we want to ensure we do not insert duplicate codes)
+
+            // dups are codes that would be removed and then immediately added again.
+            var dups = codesToAdd.Where(code => codesToRemove.Select(r => r.CodeGuid).Contains(code.CodeGuid)).ToList();
+            var codeDeletes = codesToRemove.Except(dups).ToList();
+
+            var batchInsertDtos = codesToAdd.Except(dups).Select(code => ConvertToValueSetCodeDto(valueSetGuid, code)).ToList();
+
+            var removeResult = this.PrepareRemoveCodesOperations(currentCodeDtos, codeDeletes);
+
+            var finalCodes = removeResult.CurrentCodeDtos.Union(batchInsertDtos);
+
+            return new AddRemoveCodeOperations
+            {
+                CurrentCodeDtos = currentCodeDtos,
+                NewCodeDtos = batchInsertDtos,
+                Operations = removeResult.Operations.Union(this.BuildCodeCountOperationList(valueSetGuid, finalCodes))
+                    .ToList()
+            };
+        }
+
+        private IReadOnlyCollection<Operation> BuildRemoveCodesOperationList(
+            IEnumerable<ValueSetCodeDto> originalCodeDtos,
+            IEnumerable<ICodeSystemCode> removeCodes)
+        {
+            var removeGuids = removeCodes.Select(ds => ds.CodeGuid).ToList();
+            return originalCodeDtos.Where(code => removeGuids.Contains(code.CodeGUID.GetValueOrDefault()))
+                .Select(dto => new Operation
+                {
+                    Value = dto,
+                    OperationType = OperationType.Delete
+                }).ToList();
+        }
+
+        private IReadOnlyCollection<Operation> BuildCodeCountOperationList(
+            Guid valueSetGuid,
+            IEnumerable<ValueSetCodeDto> allCodeDtos)
+        {
+            var originalCounts = this.uowManager.GetCodeCountDtos(valueSetGuid);
+            var existingCodeSystems = originalCounts.Select(c => c.CodeSystemGUID);
+
+            var allCodes = allCodeDtos as ValueSetCodeDto[] ?? allCodeDtos.ToArray();
```

Comment:
If you make `allCodes` a Lookup instead of an array, several of the following operations become simpler.

    var allCodesByCodeSystem = allCodesDtos.ToLookup(c => c.CodeSystemGuid);
    var recounts = 
        from g in allCodesByCodeSystem
        let valueSetDto = g.First()
        select new ValueSetCodeCountDto(
           valueSetDto.ValueSetGUID, 
           valueSetDto.CodeSystemGuid, // or g.Key if you prefer 
           valueSetDto.CodeSystemNM,
           g.Count()))
       .ToList();

Reply:
(you could use `GroupBy` instead of `ToLookup`: in this case there's no real difference)